modules/contact/src/userobjects/PenaltyFrictionUserObject.C

```diff
@@ -0,0 +1,351 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PenaltyFrictionUserObject.h"
+#include "MooseVariableFE.h"
+#include "SystemBase.h"
+#include "MortarUtils.h"
+#include "MooseUtils.h"
+#include "MathUtils.h"
+#include "MortarContactUtils.h"
+
+registerMooseObject("ContactApp", PenaltyFrictionUserObject);
+
+InputParameters
+PenaltyFrictionUserObject::validParams()
+{
+  InputParameters params = WeightedVelocitiesUserObject::validParams();
+  params.addClassDescription(
+      "Computes the mortar frictional contact force via a penalty approach.");
+  params.addRequiredParam<Real>("penalty", "The penalty factor for normal interaction.");
+  params.addParam<Real>("penalty_friction",
+                        "The penalty factor for frictional interaction. If not provide, the normal "
+                        "penalty factor is also used for the frictional problem.");
+  params.addRequiredParam<Real>("friction_coefficient",
+                                "The friction coefficient ruling Coulomb friction equations.");
+  return params;
+}
+
+PenaltyFrictionUserObject::PenaltyFrictionUserObject(const InputParameters & parameters)
+  : WeightedVelocitiesUserObject(parameters),
+    _penalty(getParam<Real>("penalty")),
+    _penalty_friction(isParamValid("penalty_friction") ? getParam<Real>("penalty_friction")
+                                                       : getParam<Real>("penalty")),
+    _friction_coefficient(getParam<Real>("friction_coefficient"))
+{
+  auto check_type = [this](const auto & var, const auto & var_name)
+  {
+    if (!var.isNodal())
+      paramError(var_name,
+                 "The displacement variables must have degrees of freedom exclusively on "
+                 "nodes, e.g. they should probably be of finite element type 'Lagrange'.");
+  };
+  check_type(*_disp_x_var, "disp_x");
+  check_type(*_disp_y_var, "disp_y");
+  if (_has_disp_z)
+    check_type(*_disp_z_var, "disp_z");
+}
+
+const VariableTestValue &
+PenaltyFrictionUserObject::test() const
+{
+  return _disp_x_var->phiLower();
+}
+
+const ADVariableValue &
+PenaltyFrictionUserObject::contactPressure() const
+{
+  return _contact_force;
+}
+
+const ADVariableValue &
+PenaltyFrictionUserObject::contactTangentialPressureDirOne() const
+{
+  return _frictional_contact_force_one;
+}
+
+const ADVariableValue &
+PenaltyFrictionUserObject::contactTangentialPressureDirTwo() const
+{
+  return _frictional_contact_force_two;
+}
+
+void
+PenaltyFrictionUserObject::timestepSetup()
+{
+  WeightedVelocitiesUserObject::timestepSetup();
+
+  _dof_to_old_weighted_gap.clear();
```

Comment:
So if we make `WeightedGapVelocitiesUserObject` inherit from `T`, then we wouldn't need lines like this. It would be really good I think to get to a place where derived classes only have to deal with the data they conceptually introduce. And then if someone catches a bug in penalty weighted gaps, we don't have to remember to fix both `PenaltyWeightedGap` and `PenaltyFrictionUserObject`

Reply:
In that vein, it would be cool if we could make a lot more data `private`. That would suggest we've made our design even stronger and safer!